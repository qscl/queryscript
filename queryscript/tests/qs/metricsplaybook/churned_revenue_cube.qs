import * from data;

let cte_prep =
    select
        c.id as customer_id,
        c.segment,
        c.channel,
        c.cohort,
        m.timestamp,
        m.revenue_impact,
        m.activity,
        m.plan_type
    from
        "contract_stream" m
        join "dim_customer" c
            on m.customer_id = c.id
    where
        m.activity = 'customer_churn_committed'
;

let date_slices = ['month', 'day'];
let metric_slices = ['segment', 'channel', 'plan_type'];

let cte_grouping_sets =
  select
    for slice in date_slices {
      date_trunc(slice, timestamp)::date as f"metric_{slice}",
      grouping(f"metric_{slice}") as f"{slice}_bit",
    },


    for slice in metric_slices {
      concat('{"dim_name": "', slice, '", "dim_value": "', f"{slice}", '"}') as f"combination_{slice}",
      grouping(f"combination_{slice}") as f"{slice}_bit",
    },

    'Total' as total_object,
    grouping(total_object) as total_bit,
    -- select null returns a strange type in duckdb
    -- https://github.com/qscl/queryscript/issues/73
    -- null as metric_denominators,
    'sum(revenue_impact)' as metric_calculation,
    sum(revenue_impact) as metric_value
  from
    cte_prep
  where timestamp between '2014-01-01'::timestamp and current_date() + interval 365 day
  group by grouping sets (
    for date_s in date_slices, metric_s in metric_slices {
      (f"metric_{date_s}", f"combination_{metric_s}"),
    },
    for date_s in date_slices {
      (f"metric_{date_s}", "total_object"),
    },
  )
;

SELECT COUNT(*) FROM cte_grouping_sets;

let cte_final = select
        'churned_revenue_cube' as metric_model,
        False as is_snapshot_reliant_metric,
        'timestamp' as anchor_date,
        case
          for slice in date_slices {
            when f"{slice}_bit" = 0 then slice
          }
          end as date_grain,
        case
          for slice in date_slices {
            when f"{slice}_bit" = 0 then f"metric_{slice}"
          }
          end as metric_date,
        case
          for slice in metric_slices {
            when f"{slice}_bit" = 0 then f"combination_{slice}"
          }
          end as slice_object,
        case
          for slice in metric_slices {
            when f"{slice}_bit" = 0 then concat(ifnull(json_extract_string(slice_object, '$.dim_name'), 'null'))
          }
          when total_bit = 0 then 'total'
          end as slice_dimension,
        case
          for slice in metric_slices {
            when f"{slice}_bit" = 0 then concat(ifnull(json_extract_string(slice_object, '$.dim_value'), 'null'))
          }
          when total_bit = 0 then 'Total'
          end as slice_value,
        metric_calculation,
        case
          when /* metric_denominators != 0 and */ metric_value is null then 0
          else metric_value
        end as metric_value
    from
      cte_grouping_sets;

select count(*) from cte_final;
