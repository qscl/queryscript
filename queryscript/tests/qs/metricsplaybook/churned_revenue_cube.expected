{
    "compile_errors": [
        (
            None,
            NoSuchEntry {
                path: [
                    "combination_1_bit",
                ],
                backtrace: None,
            },
        ),
        (
            None,
            NoSuchEntry {
                path: [
                    "combination_2_bit",
                ],
                backtrace: None,
            },
        ),
        (
            None,
            NoSuchEntry {
                path: [
                    "combination_3_bit",
                ],
                backtrace: None,
            },
        ),
        (
            None,
            NoSuchEntry {
                path: [
                    "combination_1",
                ],
                backtrace: None,
            },
        ),
        (
            None,
            NoSuchEntry {
                path: [
                    "combination_2",
                ],
                backtrace: None,
            },
        ),
        (
            None,
            NoSuchEntry {
                path: [
                    "combination_3",
                ],
                backtrace: None,
            },
        ),
    ],
    "decls": {
        "let contract_stream": [{
        	id Int64,
        	customer_id Int64,
        	timestamp Timestamp(
            Microsecond,
            None,
        ),
        	activity Utf8,
        	revenue_impact Float64,
        	plan_type Utf8,
        	activity_occurrence Int64,
        	activity_repeated_at Timestamp(
            Microsecond,
            None,
        ),
        }],
        "let cte_final": [?async_slot?],
        "let cte_grouping_sets": [{
        	metric_month Date32,
        	month_bit Int64,
        	metric_day Date32,
        	day_bit Int64,
        	combination_segment Utf8,
        	segment_bit Int64,
        	combination_channel Utf8,
        	channel_bit Int64,
        	combination_plan_type Utf8,
        	plan_type_bit Int64,
        	total_object Utf8,
        	total_bit Int64,
        	metric_calculation Utf8,
        	metric_value SumAgg<Float64>,
        }],
        "let cte_prep": [{
        	customer_id Int64,
        	segment Utf8,
        	channel Utf8,
        	cohort Utf8,
        	timestamp Timestamp(
            Microsecond,
            None,
        ),
        	revenue_impact Float64,
        	activity Utf8,
        	plan_type Utf8,
        }],
        "let date_slices": [Utf8],
        "let dim_customer": [{
        	id Int64,
        	segment Utf8,
        	channel Utf8,
        	first_contract_signed_date Timestamp(
            Microsecond,
            None,
        ),
        	cohort Utf8,
        }],
        "let metric_slices": [Utf8],
    },
    "queries": [
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "COUNT(*)",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| COUNT(*) |\n|----------|\n| 272      |",
            },
        ),
        Err(
            StringError {
                what: "Unknown type cannot exist at runtime (?async_slot?)",
                backtrace: None,
            },
        ),
    ],
}